tests/chainer_tests/functions_tests/connection_tests/test_embed_id.py

```diff
@@ -8,89 +8,82 @@
-        self.ggW = numpy.random.uniform(-1, 1, (3, 2)).astype(self.dtype)
-
-        self.check_backward_options = {'atol': 1e-2, 'rtol': 1e-2}
-        self.check_double_backward_options = {'atol': 1e-2, 'rtol': 1e-2}
-        if self.dtype == numpy.float16:
-            self.check_backward_options['dtype'] = numpy.float64
-            self.check_double_backward_options['dtype'] = numpy.float64
-
-    def check_forward(self, x_data, W_data):
-        x = chainer.Variable(x_data)
-        W = chainer.Variable(W_data)
-        y = chainer.functions.embed_id(x, W, self.ignore_label)
-        self.assertEqual(y.data.dtype, self.dtype)
-
-        y_expect = numpy.empty_like(self.gy)
-        for i in numpy.ndindex(self.x.shape):
-            if self.x[i] == -1:
-                y_expect[i] = 0
-            else:
-                y_expect[i] = self.W[int(self.x[i])]
-
-        testing.assert_allclose(y_expect, y.data, atol=0, rtol=0)
-
-    def test_forward_cpu(self):
-        self.check_forward(self.x, self.W)
-
-    @attr.gpu
-    def test_forward_gpu(self):
-        self.check_forward(cuda.to_gpu(self.x), cuda.to_gpu(self.W))
-
-    def check_backward(self, x_data, W_data, y_grad):
-        def f(x, W):
-            return chainer.functions.embed_id(x, W, self.ignore_label)
-
-        gradient_check.check_backward(
-            f, (x_data, W_data), y_grad, **self.check_backward_options)
-
-    def test_backward_cpu(self):
-        self.check_backward(self.x, self.W, self.gy)
-
-    @attr.gpu
-    def test_backward_gpu(self):
-        self.check_backward(
-            cuda.to_gpu(self.x), cuda.to_gpu(self.W), cuda.to_gpu(self.gy))
-
-    def check_double_backward(self, x_data, W_data, gy_data, ggW_data):
-        def f(W):
-            return chainer.functions.embed_id(
-                x_data, W, self.ignore_label)
-
-        gradient_check.check_double_backward(
-            f, W_data, gy_data, ggW_data, **self.check_double_backward_options)
-
-    def test_double_backward_cpu(self):
-        self.check_double_backward(self.x, self.W, self.gy, self.ggW)
-
-    @attr.gpu
-    def test_double_backward_gpu(self):
-        self.check_double_backward(
-            cuda.to_gpu(self.x), cuda.to_gpu(self.W), cuda.to_gpu(self.gy),
-            cuda.to_gpu(self.ggW))
+        if self.w_dtype == numpy.float16:
+            self.check_forward_options.update({
+                'rtol': 1e-2, 'atol': 1e-2})
+            self.check_backward_options.update({
+                'rtol': 1e-2, 'atol': 1e-2})
+            self.check_double_backward_options.update({
+                'rtol': 5e-3, 'atol': 5e-2})
+
+    def generate_inputs(self):
+        x = numpy.array(self.x).astype(self.x_dtype)
+        w = numpy.random.uniform(-1, 1, self.w_shape).astype(self.w_dtype)
+        return x, w,
+
+    def forward(self, inputs, device):
+        x, w = inputs
+        if self.ignore_label is None:
+            out = chainer.functions.embed_id(x, w)
+        else:
```

Comment:
why not call it directly with `self.ignore_label` instead of doing an if?

Reply:
I have made this change.

Reply:
thanks